### PR TITLE
[FLINK-36158][runtime] Refactor Job Recovery Process to Use ExecutionPlan Instead of JobGraph.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

[FLINK-36158][runtime] Refactor Job Recovery Process to Use ExecutionPlan Instead of JobGraph.

## Brief change log

[FLINK-36158][runtime] Refactor Job Recovery Process to Use ExecutionPlan Instead of JobGraph.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java [hunk 1]
```diff
@@ -109,153 +110,163 @@ public void stop() throws Exception {
...
+                        LOG.warn(
+                                "{} does not exists in {}.",
+                                executionPlan,
+                                executionPlanStateHandleStore);
                     }
                 } else {
                     throw new IllegalStateException(
-                            "Trying to update a graph you didn't "
-                                    + "#getAllSubmittedJobGraphs() or #putJobGraph() yourself before.");
+                            "Trying to update a execution plan you didn't "
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java:245]: a -> an

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java [hunk 2]
```diff
@@ -109,153 +110,163 @@ public void stop() throws Exception {
...
                 throw new NoSuchElementException(
                         String.format(
-                                "JobGraph for job [%s] was not found in JobGraphStore and is needed for attaching JobResourceRequirements.",
+                                "ExecutionPlan for job [%s] was not found in ExecutionPlanStore and is needed for attaching JobResourceRequirements.",
                                 jobId));
             }
-            JobResourceRequirements.writeToJobGraph(jobGraph, jobResourceRequirements);
-            putJobGraph(jobGraph);
+
+            checkState(executionPlan instanceof JobGraph);
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java:266]: Is this method expected to be only invoked on a `JobGraph`?

Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java:266]: Yes, the JobResourceRequirements class is designed specifically for JobGraph

**Comment by JunRuiLee** [comment]
Thanks @zhuzhurk, I've updated this pr accordingly, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java (hunk 2 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java:266]: Seems exceptions will happen if the `StreamGraph` of a streaming job is stored and `updateJobResourceRequirements(...)` is invoked.

Also even if it is a `StreamGraph`, this configuration setting can be applied via `jobConfiguration`?

Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultExecutionPlanStore.java]: Yes, you're right. I have corrected it.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request merged**